test/client/records-queue.spec.js

```diff
@@ -0,0 +1,262 @@
+                        expect( updated.item ).to.to.be.an.instanceof( Blob );
+                    }
+                } )
+                .then( done, done );
+        } );
+
+        it( 'creates a record with the current autosaved record\'s files', done => {
+            const autoSavedUpdate = Object.assign( {
+                files: files.slice(),
+            }, record );
+
+            records.updateAutoSavedRecord( autoSavedUpdate )
+                .then( () => {
+                    return records.save( 'set', record );
+                } )
+                .then( () => {
+                    return store.record.get( instanceId );
+                } )
+                .then( ( record ) => {
+                    expect( record.files.length ).to.equal( files.length );
+
+                    for ( const [ index, file ] of files.entries() ) {
+                        const updated = record.files[index];
+
+                        expect( updated.name ).to.equal( file.name );
+                        expect( updated.item ).to.to.be.an.instanceof( Blob );
+                    }
+                } )
+                .then( done, done );
+        } );
+
+        it( 'updates a record', done => {
+            const update = {
+                instanceId,
+                enketoId: 'surveyAUpdated',
+                name: 'name A updated',
+                xml: '<model><updated/></model>'
+            };
+            const payload = Object.assign( {}, update );
+
+            records.save( 'update', payload )
+                .then( () => {
+                    return store.record.get( instanceId );
+                } )
+                .then( ( record ) => {
+                    Object.entries( update ).forEach( ( [ key, value ] ) => {
+                        expect( record[key] ).to.equal( value );
+                    } );
+                } )
+                .then( done, done );
+        } );
+
+        it( 'creates a last-saved record when creating a record', done => {
+            const originalRecord = Object.assign( {}, record );
+
+            records.save( 'set', record )
+                .then( () => {
+                    return records.getLastSavedRecord();
+                } )
+                .then( ( record ) => {
+                    Object.entries( originalRecord ).forEach( ( [ key, value ] ) => {
+                        if ( key === 'instanceId' ) {
+                            expect( record[key] ).to.equal( records.getLastSavedKey() );
+                        } else if ( key === 'name' ) {
+                            expect( record[key] ).to.match( /^__lastSaved_\d+$/ );
+                        } else {
+                            expect( record[key] ).to.equal( value );
+                        }
+                    } );
+                } )
+                .then( done, done );
+        } );
+
+        it( 'replaces a last-saved record when creating a newer record', done => {
```

Comment:
I don't know if this would improve clarity. Is this because it relates to the `save` naming? My goal was to use common CRUD terms to differentiate between create/update cases.

Reply:
I guess I think of record creation as when it gets instantiated from the form definition but I see what you mean about creation really being when it gets stored. Mostly I want to make sure we're on the same page about when the last-saved record is actually created and the names in the tests don't really matter as much.